cli/validation/validate.go

```diff
@@ -0,0 +1,179 @@
+package validation
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+	"log"
+	"path/filepath"
+	"strings"
+	"github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/deployment"
+	"github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/validation/cai"
+	"github.com/forseti-security/config-validator/pkg/api/validator"
+	"github.com/forseti-security/config-validator/pkg/gcv"
+)
+
+var ancestryCache = map[string]string{}
+
+func validateAssets(assets []cai.Asset, policyRootPath string) (*validator.AuditResponse, error) {
+	return validateAssetsWithLibrary(assets,
+		filepath.Join(policyRootPath, "policies"),
+		filepath.Join(policyRootPath, "lib"))
+}
+
+func validateAssetsWithLibrary(assets []cai.Asset, policyPath, policyLibraryDir string) (*validator.AuditResponse, error) {
+	valid, err := gcv.NewValidator(
+		gcv.PolicyPath(policyPath),
+		gcv.PolicyLibraryDir(policyLibraryDir),
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing gcv validator")
+	}
+
+	pbAssets := make([]*validator.Asset, len(assets))
+	for i := range assets {
+		pbAssets[i] = &validator.Asset{}
+		if err := protoViaJSON(assets[i], pbAssets[i]); err != nil {
+			return nil, errors.Wrapf(err, "converting asset %s to proto", assets[i].Name)
+		}
+	}
+
+	if err := valid.AddData(&validator.AddDataRequest{
+		Assets: pbAssets,
+	}); err != nil {
+		return nil, errors.Wrap(err, "adding data to validator")
+	}
+
+	auditResult, err := valid.Audit(context.Background())
+	if err != nil {
+		return nil, errors.Wrap(err, "auditing")
+	}
+
+	return auditResult, nil
+}
+
+// getAncestry uses the resource manager API to get ancestry paths for
+// projects. It implements a cache because many resources share the same
+// project.
+func getAncestry(project string) (string, error) {
+	if path, ok := ancestryCache[project]; ok {
+		return path, nil
+	}
+
+	args := []string{
+		"projects",
+		"get-ancestors",
+		project,
+		"--format", "json",
+	}
+	data, err := deployment.RunGCloud(args...)
+	if err != nil {
+		return "", err
+	}
+
+	out := []struct {
+		Id   string
+		Type string
+	}{}
+	err = json.Unmarshal([]byte(data), &out)
+	if err != nil {
+		return "", err
+	}
+
+	var paths []string
+	for i := len(out) - 1; i >= 0; i-- {
+		paths = append(paths, out[i].Type+"/"+out[i].Id)
+	}
+	path := strings.Join(paths, "/")
+	ancestryCache[project] = path
+
+	return path, nil
+}
+
+func parseResourceProperties(project string, res deployment.DeploymentDescriptionResource) (cai.Resource, error) {
+	ret := cai.Resource{
+		Project: project,
+		Name:    res.Name,
+		Type:    res.Type,
+	}
+
+	props := res.Properties
```

Comment:
I'm not sure we should be unmarshaling DeploymentDescriptionResource's twice. First into DeploymentDescriptionResource, then into cai.resource.

Why don't we unmarshal the properties directly when we first unmarshal the DeploymentDescription response?

Reply:
Because properties are returned in a nested string in YAML format.